### PR TITLE
Refactor the `OTelSdkProvider` into an implementation of `OpenTelemetry`

### PR DESCRIPTION
**Pull request by cyrille-leclerc**
Refactor the `OTelSdkProvider` to make it an implementation of `OpenTelemetry` that is intended to live on a Jenkins Controller.

### Testing done

```[tasklist]
### Submitter checklist
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Review by kuisathaverat — DISMISSED** [review]

**Review by kuisathaverat — APPROVED** [review]

**Comment by cyrille-leclerc** [comment]
Sorry @kuisathaverat I made some changes to better support OTel instrumentation in other plugins. Could you please review again the PR 👼 
Note that I have added many unit tests with the last commit

File: src/main/java/io/jenkins/plugins/opentelemetry/opentelemetry/InstrumentationScope.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+public class InstrumentationScope {
+    @Nonnull
+    final String instrumentationScopeName;
+    @Nullable
+    final String schemaUrl;
+    @Nullable
+    final String instrumentationScopeVersion;
```
Comment (kuisathaverat) [inline src/main/java/io/jenkins/plugins/opentelemetry/opentelemetry/InstrumentationScope.java:18]: I think it is worth adding a JavaDoc description of the fields and the class

File: src/main/java/io/jenkins/plugins/opentelemetry/opentelemetry/ReconfigurableLoggerProvider.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import io.opentelemetry.api.logs.LoggerBuilder;
+import io.opentelemetry.api.logs.LoggerProvider;
+import io.opentelemetry.api.trace.TracerBuilder;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class ReconfigurableLoggerProvider implements LoggerProvider {
```
Comment (kuisathaverat) [inline src/main/java/io/jenkins/plugins/opentelemetry/opentelemetry/ReconfigurableLoggerProvider.java:20]: I think it is worth adding a JavaDoc description to the class

File: src/main/java/io/jenkins/plugins/opentelemetry/opentelemetry/ReconfigurableOpenTelemetry.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * Reconfigurable {@link OpenTelemetry}
+ */
+public class ReconfigurableOpenTelemetry implements OpenTelemetry, Closeable {
```
Comment (kuisathaverat) [inline src/main/java/io/jenkins/plugins/opentelemetry/opentelemetry/ReconfigurableOpenTelemetry.java:46]: I think it is worth adding a JavaDoc description to the class

File: src/main/java/io/jenkins/plugins/opentelemetry/opentelemetry/ReconfigurableTracerProvider.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+import io.opentelemetry.api.trace.TracerBuilder;
+import io.opentelemetry.api.trace.TracerProvider;
+
+import javax.annotation.Nonnull;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class ReconfigurableTracerProvider implements TracerProvider {
```
Comment (kuisathaverat) [inline src/main/java/io/jenkins/plugins/opentelemetry/opentelemetry/ReconfigurableTracerProvider.java:20]: I think it is worth adding a JavaDoc description to the class

**Review by kuisathaverat — APPROVED** [review]
minor comments LGTM

**Comment by cyrille-leclerc** [comment]
Thanks, I'll javadoc in a subsequent PR

File: src/main/java/io/jenkins/plugins/opentelemetry/JenkinsOpenTelemetryPluginConfiguration.java [hunk 5]
```diff
@@ -218,19 +217,33 @@ public OpenTelemetryConfiguration toOpenTelemetryConfiguration() {
...
 
+    /**
+     * Register reconfigurable {@link io.opentelemetry.api.OpenTelemetry}
+     * on {@link  io.opentelemetry.api.GlobalOpenTelemetry}
+     * and {@link io.jenkins.plugins.opentelemetry.opentelemetry.ReconfigurableEventLoggerProvider}
+     * on {@link io.opentelemetry.api.incubator.events.GlobalEventLoggerProvider}
+     * as early as possible in Jenkins lifecycle so any plugin invoking those Global setters will have the
+     * reconfigurable instance .
+     */
+    @Initializer(after = InitMilestone.EXTENSIONS_AUGMENTED, before = InitMilestone.SYSTEM_CONFIG_LOADED)
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/opentelemetry/JenkinsOpenTelemetryPluginConfiguration.java:228]: before system config loaded? how will the configuration be available for this plugin? and how will someone configure this plugin with JCasC if its so early?

jcasc is after system config loaded:
https://github.com/jenkinsci/configuration-as-code-plugin/blob/3b0e6de4ab83e8024e6cac635728c28b68ac663a/plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java#L337

Comment (cyrille-leclerc) [reply src/main/java/io/jenkins/plugins/opentelemetry/JenkinsOpenTelemetryPluginConfiguration.java:228]: Correct: There reasons is:
1. we want to be sure that any plugin that is using OTel will use our implementation of `OpenTelemetry`, even when getting this OpenTelemetry instance through `GlobalOpenTelemetry.get()`
2. we don't know how to ensure the OpenTelemetry plugin gets configured & initialized before all other plugins (e.g. the initialization code [`DatabaseSchemaLoader#migrateSchema()`](https://github.com/jenkinsci/junit-sql-storage-plugin/blob/76cefd5f4cf6117553fa6f219442a9cbba64b458/src/main/java/io/jenkins/plugins/junit/storage/database/DatabaseSchemaLoader.java#L24-L26) of the JUnit SQL Storage Plugin is invoked by Jenkins before the Jenkins OTel Plugin initialization code [`JenkinsOpenTelemetryPluginConfiguration#initializeOpenTelemetry()`](https://github.com/jenkinsci/opentelemetry-plugin/blob/3aa8270a1fd83c8b53014ac66c448bb0b33108fd/src/main/java/io/jenkins/plugins/opentelemetry/JenkinsOpenTelemetryPluginConfiguration.java#L242C17-L242C40) despite the dependency.

The solution we identified is to:
1. Register a Reconfigurable OpenTelemetry instance in `GlobalOpenTelemetry` as early as possible in Jenkin's lifecycle with a NoOp implementation (see `initializeOpenTelemetryAfterExtensionsAugmented()`).
2. When the Jenkins Otel Plugin config is ready, we reconfigure the OpenTelmetry instance (see `initializeOpenTelemetry()`)

I think we can cleanup a bit this code but it works and it was quite time consuming to get there :-) 

Dos it make sense?

https://github.com/jenkinsci/junit-sql-storage-plugin/blob/76cefd5f4cf6117553fa6f219442a9cbba64b458/src/main/java/io/jenkins/plugins/junit/storage/database/DatabaseSchemaLoader.java#L24-L26

Comment (timja) [reply src/main/java/io/jenkins/plugins/opentelemetry/JenkinsOpenTelemetryPluginConfiguration.java:228]: ah I see bit confusing but it'll work I think

**Pull request merged**